File: leet_code/sort_search/kth_largest_in_array.js
// Find the kth largest element in an unsorted array. Note that it is the kth largest element in the sorted order, not the kth distinct element.
//
// Example 1:
//
// Input: [3,2,1,5,6,4] and k = 2
// Output: 5

const quicksort = arr => {
  if (arr.length <= 1) return arr;

  const pivot = arr[0],
        left = [],
        right = [];

  arr.slice(1).forEach((el) => {
    el < pivot ? left.push(el) : right.push(el);
  });

  return quicksort(left).concat([pivot]).concat(quicksort(right));
};

const findKthLargest = (nums, k) => {
    return quicksort(nums)[nums.length - k];
};

File: leet_code/sort_search/kth_largest_in_array.rb
# Find the kth largest element in an unsorted array. Note that it is the kth largest element in the sorted order, not the kth distinct element.
#
# Example 1:
#
# Input: [3,2,1,5,6,4] and k = 2
# Output: 5

def quicksort(arr)
  return arr if arr.length <= 1

  pivot, left, right = arr[0], [], []

  arr[1..-1].each do |el|
    el < pivot ? left << el : right << el
  end
  quicksort(left).concat([pivot]).concat(right)
end

def find_kth_largest(nums, k)
  quicksort(nums)[nums.length - k]
end
